{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "29def768",
   "metadata": {
    "id": "29def768"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/clami66/AF_unmasked/blob/notebook/notebooks/AF_unmasked.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6KgI44cIYON",
   "metadata": {
    "id": "d6KgI44cIYON"
   },
   "source": [
    "# <font color='#2684ff'>AF_unmasked: a simplified notebook</font>\n",
    "\n",
    "<img src=\"https://github.com/clami66/AF_unmasked/raw/main/fig/header.png\" height=\"200\">\n",
    "\n",
    "This notebook allows to run AF_unmasked on multimeric sequences and templates of your choice. Not all features implemented on the command line version of AF_unmasked are currently available on the notebook, but more will come later.\n",
    "\n",
    "This version of AF_unmasked relies on MMseqs2 alignments, run by the [ColabFold](https://github.com/sokrypton/ColabFold) MSA server. Some of the code on this notebook is also based or taken from the ColabFold notebook.\n",
    "\n",
    "<font color='#2684ff'>**References**</font>\n",
    "\n",
    "If you use this version of AF_unmasked in your research, consider citing:\n",
    "\n",
    "- Mirabello et al.: \"Unmasking AlphaFold to integrate experiments and predictions in multimeric complexes\". [Nature Communications volume 15, 8724 (2024)](https://www.nature.com/articles/s41467-024-52951-w)\n",
    "- Jumper et al.: \"Highly accurate protein structure prediction with AlphaFold\". [Nature volume 596, pages 583–589 (2021)](https://www.nature.com/articles/s41586-021-03819-2)\n",
    "- Evans et al.: \"Protein complex prediction with AlphaFold-Multimer\". [BiorXiv](https://www.biorxiv.org/content/early/2021/10/04/2021.10.04.463034)\n",
    "- Mirdita et al. \"ColabFold: making protein folding accessible to all\" [Nature Methods volume 19, pages 679–682 (2022)](https://www.nature.com/articles/s41592-022-01488-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d0399a",
   "metadata": {},
   "source": [
    "## <font color='#2684ff'>Instructions</font>\n",
    "\n",
    "The job is configured and launched from the side menu on the left. \n",
    "\n",
    "1. <font color='#2684ff'> **Select a name for your job**</font>. Confirm by clicking on the button that will appear or press enter.\n",
    "1. <font color='#2684ff'> **Upload files**</font>. Two files are neede to run the job:\n",
    "    * a FASTA file containing multiple sequences for a multimer to predict\n",
    "    * a PDB (.pdb or .cif) file containing template information for each chain in the multimer\n",
    "1. <font color='#2684ff'> **Configure the MSA/templates**</font>:\n",
    "    * **MSA mode**. MSA is disabled by default. Switch to mmseqs2_uniref when inpainting, use mmseqs2_uniref_env if you want to increase MSA depth.\n",
    "    * **Number of sequences in MSA**. Increase if the inpainted areas are badly predicted, decrease if the prediction does not resemble the template.\n",
    "    * **Automatically inpaint clashes**. We recommend leaving this on.\n",
    "    * **Align template sequences to targets**. Disable to let AlphaFold handle the target-template alignments. This is better with remote homologs, but might not work with homomers.\n",
    "    * **Target-template mapping**. A dropdown menu will appear for each sequence in the fasta file. Select the appropriate chain from the PDB template so that AF_unmasked can use the correct target-template mapping.\n",
    "1. <font color='#2684ff'> **Format templates**</font>. Press the blue button to convert the templates. Alignments between target and template sequences will be shown in the output window. Make sure the alignments are correct before running the following step.   \n",
    "1. <font color='#2684ff'> **Configure AlphaFold settings**</font>:\n",
    "    * **Select AF model type (v2 or v3)**. v2 tends to work better with multimeric templates, but v3 might be better with large multimers. The server generates only model 5 predictions (model_5_v2 or model_5_v3)\n",
    "    * **Number of predictions**. We recommend starting with 1 until you're happy with the options you have selected\n",
    "    * **Number of recycles**. Few recycles will be faster. Increasing the recycles might be useful if you are inpainting parts of the structure.\n",
    "    * **Recycle early stop tolerance**. Leave as default unless you want to run all recycles. In that case lower to 0.0\n",
    "    * **Use dropout**. Disable if you generate few samples, enable if you generate a lot of samples and want some more diversity in the predictions\n",
    "\n",
    "1. <font color='#2684ff'> **Run the AF_unmasked job**</font> by clicking on the blue button once you're happy with the parameters and target-template mapping\n",
    "\n",
    "1. <font color='#2684ff'> **Check and download outputs**</font>. Results will appear in this window. To download the predictions, click on \"Output files\" at the bottom of the side menu. You will have the option to include all pickle (.pkl) files including extra outputs if needed, keep in mind that these tend to be quite large.\n",
    "    \n",
    "## <font color='#2684ff'>Outputs</font>\n",
    "\n",
    "Outputs will appear below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a23d00d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "is_test = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64fadb4d-4e38-4967-9e78-fb980eef13cf",
   "metadata": {
    "cellView": "form",
    "id": "64fadb4d-4e38-4967-9e78-fb980eef13cf"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import pickle\n",
    "import shutil\n",
    "import importlib_metadata\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from string import ascii_uppercase, ascii_lowercase\n",
    "\n",
    "\n",
    "class color:\n",
    "   PURPLE = '\\033[95m'\n",
    "   CYAN = '\\033[96m'\n",
    "   DARKCYAN = '\\033[36m'\n",
    "   BLUE = '\\033[94m'\n",
    "   GREEN = '\\033[92m'\n",
    "   YELLOW = '\\033[93m'\n",
    "   RED = '\\033[91m'\n",
    "   BOLD = '\\033[1m'\n",
    "   UNDERLINE = '\\033[4m'\n",
    "   END = '\\033[0m'\n",
    "\n",
    "cwd = Path.cwd()\n",
    "ascii_upperlower = ascii_uppercase + ascii_lowercase\n",
    "\n",
    "python_version = f\"{sys.version_info.major}.{sys.version_info.minor}\"\n",
    "PYTHON_VERSION = python_version\n",
    "\n",
    "print(color.BOLD + \"Setting up the environment... \", end=\"\")\n",
    "\n",
    "if not os.path.isfile(\"COLABFOLD_READY\"):\n",
    "  print(\"Installing dependencies...\")\n",
    "  os.system(\"git clone -b notebook https://github.com/clami66/AF_unmasked.git\")\n",
    "  os.system(\"pip install numpy<1.20 biopython==1.79 'colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold'\")\n",
    "  \n",
    "  os.system(\"mamba install -y -c conda-forge -c bioconda hmmer kalign2=2.04 hhsuite=3.3.0\")\n",
    "  os.system(\"touch COLABFOLD_READY\")\n",
    "\n",
    "sys.path.insert(0, f\"{cwd}/AF_unmasked\")\n",
    "sys.path.insert(0, f\"{cwd}\")\n",
    "\n",
    "import mercury as mr\n",
    "\n",
    "from AF_unmasked.alphafold.data.prepare_templates import *\n",
    "from AF_unmasked.alphafold.data.mmseqs_2_uniprot import *\n",
    "from Bio import Align, SeqIO, AlignIO\n",
    "from Bio.PDB.mmcifio import MMCIFIO\n",
    "from Bio.PDB import MMCIFParser\n",
    "\n",
    "from colabfold.batch import get_msa_and_templates, msa_to_str\n",
    "from colabfold.utils import DEFAULT_API_SERVER, get_commit\n",
    "from colabfold.download import download_alphafold_params\n",
    "\n",
    "from AF_unmasked.run_alphafold import predict_structure\n",
    "from AF_unmasked.alphafold.data import pipeline, pipeline_multimer\n",
    "\n",
    "from AF_unmasked.alphafold.data.tools import hmmsearch, jackhmmer\n",
    "from AF_unmasked.alphafold.data import templates\n",
    "\n",
    "from AF_unmasked.alphafold.model import model, data, config\n",
    "\n",
    "import py3Dmol\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.image as mpimg\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.special import softmax\n",
    "\n",
    "from colabfold.colabfold import plot_plddt_legend\n",
    "from colabfold.colabfold import pymol_color_list, alphabet_list\n",
    "\n",
    "print(\"Done.\"  + color.END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b4b61f3-bfeb-414e-8962-fc1cb7855b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "app = mr.App(title=\"AF_unmasked\",\n",
    "        description=\"\",\n",
    "        show_code=False,\n",
    "        show_prompt=False,\n",
    "        continuous_update=True,\n",
    "        static_notebook=False,\n",
    "        show_sidebar=True,\n",
    "        full_screen=True,\n",
    "        allow_download=False,\n",
    "        allow_share=True,\n",
    "        stop_on_error=True\n",
    ")\n",
    "\n",
    "jobname = mr.Text(value=\"\", label=\"Job name\", rows=1)\n",
    "if not jobname.value:\n",
    "    mr.Stop()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae62d219-2628-421d-a646-05261cd67a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = mr.OutputDir()\n",
    "out_dir = Path(outdir.path, jobname.value)\n",
    "out_dir.mkdir(parents=True, exist_ok=True)\n",
    "targets = out_dir.joinpath(f\"{jobname.value}.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0962e716-b4f4-4177-874f-5042ad413259",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add file upload widget\n",
    "f = mr.File(label=\"Upload a single .fasta file for a multimeric target\", max_file_size=\"1MB\")\n",
    "fasta = f.filepath\n",
    "p = mr.File(label=\"Upload a template PDB (.pdb, .cif)\", max_file_size=\"20MB\")\n",
    "pdb = p.filepath\n",
    "\n",
    "if not fasta or not pdb:\n",
    "    mr.Stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62803f6f-4e3a-4823-98a1-a7fe24f1c43c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if p.filename.endswith(\".pdb\"):\n",
    "  template_format = \"pdb\"\n",
    "elif p.filename.endswith(\".cif\"):\n",
    "  template_format = \"cif\"\n",
    "else:\n",
    "  raise ValueError(\"Template must be in .pdb or .cif format\")\n",
    "\n",
    "fasta = Path(fasta)\n",
    "pdb = Path(pdb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9c59f23",
   "metadata": {},
   "source": [
    "### <font color='#2684ff'>Sequences and alignments</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fe53b5e-e8a1-43b0-91c0-4a9c55e32499",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = shutil.copyfile(fasta, targets)\n",
    "\n",
    "if not is_fasta(targets):\n",
    "  raise ValueError(\"\"\"The input does not appear to be in fasta format\n",
    "  Example of fasta format input:\n",
    "  > H1142_A\n",
    "  GLEKDFLPLYFGWFLTK...\n",
    "  > H1142_B\n",
    "  EVQLEESGGGLVQAGGS...\n",
    "  \"\"\")\n",
    "\n",
    "with open(targets, \"r\") as f:\n",
    "  print(color.BOLD + color.RED + \"Fasta sequences:\" + color.END)\n",
    "  print(f.read())\n",
    "  #print()\n",
    "\n",
    "seq2chain = {}\n",
    "chain_idx = 0\n",
    "for record in SeqIO.parse(targets, \"fasta\"):\n",
    "  if record.seq not in seq2chain:\n",
    "    seq2chain[record.seq] = [ascii_upperlower[chain_idx]]\n",
    "  else:\n",
    "    seq2chain[record.seq].append(ascii_upperlower[chain_idx])\n",
    "  chain_idx += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04b0158a-3425-403b-8e49-4a61b99d9fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = out_dir.joinpath(p.filename)\n",
    "_ = shutil.copyfile(pdb, template)\n",
    "\n",
    "# template data\n",
    "template_model = load_PDB(template, is_mmcif=(template_format == \"cif\"))\n",
    "template_chains = [c.id for c in template_model]\n",
    "remove_extra_chains(template_model, template_chains)\n",
    "remove_hetatms(template_model)\n",
    "template_sequences = [\n",
    "        get_fastaseq(template_model, chain) for chain in template_chains\n",
    "    ]\n",
    "\n",
    "print(color.BOLD + color.RED + \"Template sequences:\" + color.END)\n",
    "for seq, chain in zip(template_sequences, template_chains):\n",
    "  print(f\"Chain {chain}: {seq}\")\n",
    "\n",
    "# target data\n",
    "target_chains, target_sequences, target_models = get_target_data(\n",
    "            [str(targets)],\n",
    "            chains=None,\n",
    "            is_fasta=True,\n",
    "            is_mmcif=False,\n",
    "        )\n",
    "assert len(target_chains) <= len(\n",
    "      template_chains\n",
    "), f\"Not enough chains in the template structure to cover all target chains. Partial templates are currently not supported on the colab version of AF_unmasked.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca11f0b3-fbec-4015-b89e-2b49b1d17e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MSA and template options\n",
    "msa_mode = mr.Select(label=\"MSA mode\", choices=[\"no_MSA\", \"mmseqs2_uniref\", \"mmseqs2_uniref_env\"])\n",
    "\n",
    "msa_depth = mr.Select(label=\"Number of sequences in MSA\",\n",
    "                          choices=[\"auto\", \"1\", \"32\", \"64\", \"256\", \"512\"])\n",
    "\n",
    "msa_depth = None if msa_depth.value == \"auto\" else int(msa_depth.value)\n",
    "\n",
    "inpaint_clashes = mr.Checkbox(value=True, label=\"Automatically inpaint clashes in template\")\n",
    "align = mr.Checkbox(value=True, label=\"Align template sequences to targets\")\n",
    "\n",
    "align_tool = \"blast\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f79e2f5-a679-4c09-90d9-ed4fe92376b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_preview = [f\"Chain {template_chain} (seq: {template_seq[:10]}...)\" for template_chain, template_seq in zip(template_chains, template_sequences)]\n",
    "\n",
    "template_c = []\n",
    "for i, ch in enumerate(target_chains):\n",
    "    template_c.append(mr.Select(label=f\"Select template chain for fasta sequence {i+1} (seq: {target_sequences[i][:10]}\", value=template_preview[i], choices=template_preview, url_key=i))\n",
    "\n",
    "repeat_template = mr.Select(label=\"Repeat template\", choices=[\"1 time\", \"2 times\", \"3 times\", \"4 times\"], value=\"1 time\")\n",
    "\n",
    "run_temp = mr.Button(label=\"Format templates\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e06af2-b2eb-41e2-805b-89819a52de47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data pipeline\n",
    "\n",
    "\n",
    "done = False\n",
    "\n",
    "if run_temp.clicked:\n",
    "    temp_reps = int(repeat_template.value.split()[0])\n",
    "    template_chains = [temp.value.split()[1] for temp in template_c]\n",
    "    \n",
    "    if len(template_chains) != len(set(template_chains)):\n",
    "      raise ValueError(\"Must select a different template chain for each fasta sequence\")\n",
    "    \n",
    "    append = False\n",
    "    \n",
    "    mmcif_path = Path(out_dir, \"template_data\", \"mmcif_files\")\n",
    "    mmcif_path.mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    for i in range(temp_reps):\n",
    "        print(color.BOLD + f\"Filling template slot n. {i+1}...\" + color.END)\n",
    "        next_id = get_next_id(mmcif_path) if append else \"0000\"\n",
    "    \n",
    "        io = MMCIFIO()\n",
    "        template_mmcif_path = os.path.join(\n",
    "            out_dir, \"template_data\", \"mmcif_files\", f\"{next_id}.cif\"\n",
    "        )\n",
    "    \n",
    "        if inpaint_clashes.value:\n",
    "            template_model = detect_and_remove_clashes(template_model)\n",
    "            template_sequences = [\n",
    "                get_fastaseq(template_model, chain) for chain in template_chains\n",
    "            ]\n",
    "    \n",
    "        io.set_structure(template_model)\n",
    "        io.save(template_mmcif_path)\n",
    "    \n",
    "        fix_mmcif(\n",
    "            template_mmcif_path, template_chains, template_sequences, \"2100-01-01\"\n",
    "        )\n",
    "    \n",
    "        pdb_seqres_path = Path(out_dir, \"template_data\", \"pdb_seqres.txt\").resolve()\n",
    "        write_seqres(\n",
    "            pdb_seqres_path,\n",
    "            template_sequences,\n",
    "            template_chains,\n",
    "            seq_id=next_id,\n",
    "            append=append,\n",
    "        )\n",
    "    \n",
    "        # extra flagfile for AF usage\n",
    "        af_flagfile_path = Path(out_dir, \"template_data\", \"templates.flag\")\n",
    "        if i == 0:  # don't overwrite file if already there\n",
    "            with open(af_flagfile_path, \"w\") as flagfile:\n",
    "                flagfile.write(f\"--template_mmcif_dir={mmcif_path.resolve()}\\n\")\n",
    "                flagfile.write(f\"--pdb_seqres_database_path={pdb_seqres_path}\\n\")\n",
    "                if align.value:  # means we are not going to let AF overwrite pdb_hits.sto\n",
    "                    flagfile.write(\"--use_precomputed_msas\\n\")\n",
    "    \n",
    "        if True:\n",
    "    \n",
    "            assert len(target_chains) == len(\n",
    "                template_chains\n",
    "            ), f\"The number of chains to align from target ({target_chains}) doesn't match the number of chains in the template ({template_chains}). Make sure that the files contain the same number of chains or select the chains that should be paired with --target_chains, --template_chains\"\n",
    "            for (\n",
    "                i,\n",
    "                (\n",
    "                    template_chain,\n",
    "                    template_sequence,\n",
    "                    target_chain,\n",
    "                    target_sequence,\n",
    "                    target_model,\n",
    "                ),\n",
    "            ) in enumerate(\n",
    "                zip(\n",
    "                    template_chains,\n",
    "                    template_sequences,\n",
    "                    target_chains,\n",
    "                    target_sequences,\n",
    "                    target_models,\n",
    "                )\n",
    "            ):\n",
    "                msa_chain = ascii_upperlower[i]\n",
    "                this_template_model = pickle.loads(pickle.dumps(template_model, -1))\n",
    "                this_target_model = pickle.loads(pickle.dumps(target_model, -1))\n",
    "                print(color.BOLD + color.RED + f\"Aligning fasta sequence {i+1} (seq: {target_sequence[0:10]}...) to template chain {template_chain} (seq: {template_sequence[0:10]}...)\" + color.END)\n",
    "                alignment = do_align(\n",
    "                    template_sequence,\n",
    "                    this_template_model,\n",
    "                    target_sequence,\n",
    "                    this_target_model,\n",
    "                    alignment_type=\"blast\",\n",
    "                )\n",
    "                sto_alignment = format_alignment_stockholm(\n",
    "                    alignment, hit_id=next_id, hit_chain=template_chain\n",
    "                )\n",
    "    \n",
    "    \n",
    "                msa_path = f\"msas/{msa_chain}\"\n",
    "    \n",
    "                # write alignment to file\n",
    "                Path(out_dir, msa_path).mkdir(parents=True, exist_ok=True)\n",
    "                with open(\n",
    "                    Path(out_dir, msa_path, \"pdb_hits.sto\"),\n",
    "                    mode=\"a\" if append else \"w\",\n",
    "                ) as pdb_hits:\n",
    "                    for line in sto_alignment:\n",
    "                        pdb_hits.write(line)\n",
    "        if temp_reps > 1:\n",
    "            append = True\n",
    "\n",
    "    if msa_mode.value == \"no_MSA\": # same as \"no_MSA\" on the AF_unmasked paper\n",
    "      unpaired_msa = []\n",
    "      for i, ts in enumerate(target_sequences):\n",
    "        unpaired_msa.append(f\"> seq_{i}\\n{ts}\")\n",
    "      query_seqs_unique = set(target_sequences)\n",
    "    else: # Alignments rely on ColabFold's MSA server\n",
    "      print(\"Querying ColabFold's MSA server\")\n",
    "      msa_lines = None\n",
    "      use_templates = False\n",
    "      custom_template_path = None\n",
    "      pair_mode = \"unpaired\"\n",
    "      pairing_strategy = \"greedy\"\n",
    "      host_url = DEFAULT_API_SERVER\n",
    "      version = importlib_metadata.version(\"colabfold\")\n",
    "      commit = get_commit()\n",
    "      if commit:\n",
    "          version += f\" ({commit})\"\n",
    "      user_agent = f\"colabfold/{version}\"\n",
    "\n",
    "      unpaired_msa, paired_msa, query_seqs_unique, query_seqs_cardinality, template_features = get_msa_and_templates(jobname.value, target_sequences, msa_lines, out_dir, msa_mode, use_templates,\n",
    "                              custom_template_path, pair_mode, pairing_strategy, host_url, user_agent)\n",
    "\n",
    "    for sequence, msa in zip(query_seqs_unique, unpaired_msa):\n",
    "      chains = seq2chain[sequence]\n",
    "      for chain in chains:\n",
    "        out_dir.joinpath(f\"msas/{chain}/bfd_uniref_hits.a3m\").write_text(msa)\n",
    "        pseudo_uniprot = open(out_dir.joinpath(f\"msas/{chain}/uniprot_hits.a3m\"), \"w\")\n",
    "        pseudo_uniprot.write(f\"> {chain}\\n\")\n",
    "        pseudo_uniprot.write(str(sequence))\n",
    "        pseudo_uniprot.close()\n",
    "\n",
    "        input_handle  = open(out_dir.joinpath(f\"msas/{chain}/uniprot_hits.a3m\"), \"r\")\n",
    "        output_handle = open(out_dir.joinpath(f\"msas/{chain}/uniprot_hits.sto\"), \"w\")\n",
    "\n",
    "        alignments = AlignIO.parse(input_handle, \"fasta\")\n",
    "        AlignIO.write(alignments, output_handle, \"stockholm\")\n",
    "\n",
    "        output_handle.close()\n",
    "        input_handle.close()\n",
    "\n",
    "    template_searcher = hmmsearch.Hmmsearch(\n",
    "        binary_path=shutil.which(\"hmmsearch\"),\n",
    "        hmmbuild_binary_path=shutil.which(\"hmmbuild\"),\n",
    "        database_path=out_dir.joinpath(f\"template_data/pdb_seqres.txt\"))\n",
    "\n",
    "    template_featurizer = templates.HmmsearchHitFeaturizer(\n",
    "        mmcif_dir=mmcif_path.resolve(),\n",
    "        max_template_date=\"2100-01-01\",\n",
    "        max_hits=4,\n",
    "        kalign_binary_path=shutil.which(\"kalign\"),\n",
    "        release_dates_path=None,\n",
    "        obsolete_pdbs_path=None)\n",
    "\n",
    "    monomer_data_pipeline = pipeline.DataPipeline(\n",
    "        jackhmmer_binary_path=shutil.which(\"jackhmmer\"),\n",
    "        hhblits_binary_path=shutil.which(\"hhblits\"),\n",
    "        uniref90_database_path=\".\",\n",
    "        mgnify_database_path=\"\",\n",
    "        bfd_database_path=\"\",\n",
    "        uniref30_database_path=\"\",\n",
    "        small_bfd_database_path=\"\",\n",
    "        template_searcher=template_searcher,\n",
    "        template_featurizer=template_featurizer,\n",
    "        use_small_bfd=False,\n",
    "        use_precomputed_msas=True,\n",
    "        mgnify_max_hits=1,\n",
    "        uniref_max_hits=1,\n",
    "        bfd_max_hits=msa_depth,\n",
    "        no_uniref=True,\n",
    "        no_mgnify=True)\n",
    "\n",
    "    data_pipeline = pipeline_multimer.DataPipeline(\n",
    "        monomer_data_pipeline=monomer_data_pipeline,\n",
    "        jackhmmer_binary_path=shutil.which(\"jackhmmer\"),\n",
    "        uniprot_database_path=None,\n",
    "        use_precomputed_msas=True,\n",
    "        max_uniprot_hits=1,\n",
    "        separate_homomer_msas=True)\n",
    "\n",
    "\n",
    "    done = True\n",
    "\n",
    "if not done:\n",
    "    mr.Stop()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e38b756e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NN options\n",
    "model_type = mr.Select(label=\"Select AF model type (v2 or v3)\",\n",
    "                          choices=[\"alphafold2_multimer_v2\", \"alphafold2_multimer_v3\"])\n",
    "\n",
    "predictions_per_model = mr.Select(choices=[\"1\", \"2\", \"3\", \"5\", \"10\", \"20\"], value=\"1\", label=\"Number of predictions\")\n",
    "n_preds = predictions_per_model.value\n",
    "\n",
    "num_recycles = mr.Select(choices=[\"0\", \"1\", \"2\", \"3\", \"12\", \"20\"], value=\"3\", label=\"Number of recycles\")\n",
    "n_recy = num_recycles.value\n",
    "recycle_early_stop_tolerance = 0.5 #mr.Numeric(value=0.5, min=0.0, max=1.0, label=\"Recycle early stop tolerance\", step=1)\n",
    "\n",
    "use_dropout = mr.Checkbox(value=True, label=\"Use dropout (increases sampling noise)\")\n",
    "drop = use_dropout.value\n",
    "\n",
    "run_pred = mr.Button(label=\"Run AF_unmasked\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d656c6d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prediction step\n",
    "\n",
    "\n",
    "done2 = False\n",
    "if run_pred.clicked:\n",
    "    data_dir = Path(\"./\")\n",
    "    if not glob.glob(f\"{data_dir}/params/*_finished.txt\"):\n",
    "      print(\"downloading AF parameters...\")\n",
    "      download_alphafold_params(model_type.value, data_dir)\n",
    "\n",
    "    model_names = [\"model_5_multimer_v2\"] if model_type.value == \"alphafold2_multimer_v2\" else [\"model_5_multimer_v3\"]\n",
    "    model_runners = {}\n",
    "\n",
    "    for model_name in model_names:\n",
    "        model_config = config.model_config(model_name)\n",
    "        model_config.model.num_ensemble_eval = 1\n",
    "        model_config.model.embeddings_and_evoformer.cross_chain_templates = True\n",
    "        model_config.model.num_recycle = int(n_recy)\n",
    "        model_config.model.global_config.eval_dropout = drop\n",
    "        model_config.model.recycle_early_stop_tolerance = recycle_early_stop_tolerance\n",
    "\n",
    "        model_params = data.get_model_haiku_params(\n",
    "            model_name=model_name, data_dir=str(data_dir))\n",
    "        model_runner = model.RunModel(model_config, model_params)\n",
    "        for i in range(int(n_preds)):\n",
    "          model_runners[f'{model_name}_pred_{i}'] = model_runner\n",
    "\n",
    "    if not is_test:\n",
    "        predict_structure(\n",
    "            fasta_path=targets,\n",
    "            fasta_name=jobname.value,\n",
    "            output_dir_base=f\"{outdir.path}\",\n",
    "            data_pipeline=data_pipeline,\n",
    "            model_runners=model_runners,\n",
    "            benchmark=False,\n",
    "            random_seed=0,\n",
    "            models_to_relax=None)\n",
    "    else:\n",
    "        template =  os.path.join(out_dir, \"template_data\", \"mmcif_files\", \"0000.cif\")\n",
    "        os.system(f\"rsync -av ./mercury_example/ {out_dir}\")\n",
    "    done2 = True\n",
    "\n",
    "if not done2:\n",
    "    mr.Stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b55d3b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude_pickles = True \n",
    "# mr.Checkbox(value=True, label=\"Exclude pickle files (.pkl) from download\")\n",
    "# run_zip = mr.Button(label=\"Zip results for download\")\n",
    "#if run_zip.clicked:\n",
    "extra_zip_flags = \"-x '*.pkl'\" if exclude_pickles else \"\"\n",
    "results_zip = f\"{jobname.value}.zip\"\n",
    "_ = os.system(f\"cd {outdir.path} && zip {extra_zip_flags} -r {results_zip} {jobname.value}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d792d16b",
   "metadata": {},
   "source": [
    "### <font color='#2684ff'>Scores and plots </font>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6973f374",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking_debug = out_dir.joinpath(\"ranking_debug.json\")\n",
    "rank_choices = [Path(ranked_pdb).stem for ranked_pdb in glob.glob(f\"{out_dir}/ranked_*.pdb\")]\n",
    "rank_choices = sorted(rank_choices, key=lambda x: int(x.split(\"_\")[-1].split(\".\")[0]))\n",
    "if ranking_debug.is_file():\n",
    "    print(color.BOLD + color.RED + \"Ranking confidence scores:\" + color.END)\n",
    "    ranking = json.load(open(ranking_debug))\n",
    "    for i, key in enumerate(ranking[\"order\"]):\n",
    "        if \"iptm+ptm\" in ranking:\n",
    "            print(f\"  ranked_{i}: {ranking['iptm+ptm'][key]}\")\n",
    "        else:\n",
    "            print(f\"  ranked_{i}: {ranking['plddts'][key]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3601c693",
   "metadata": {},
   "outputs": [],
   "source": [
    "if msa_mode.value != \"no_MSA\":\n",
    "    print(color.BOLD + color.RED + \"MSA:\" + color.END)\n",
    "    msa = pickle.load(open(f\"{out_dir}/features.pkl\", \"rb\"))[\"msa\"]\n",
    "    plt.imshow(msa)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda03d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranked_pdb = mr.Select(label=\"Select structure by rank\", choices=rank_choices, value=rank_choices[0])\n",
    "col = mr.Select(label=\"Select structure coloring\", choices=[\"chain\", \"lDDT\", \"rainbow\"], value=\"chain\")\n",
    "show_sidechains = mr.Checkbox(value=False, label=\"Show sidechains\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fd51f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_pdb(pdb_file, extension, show_sidechains=False, show_mainchains=False, color=\"lDDT\"):\n",
    "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',)\n",
    "  view.addModel(open(pdb_file,'r').read(), extension)\n",
    "\n",
    "  if color == \"lDDT\":\n",
    "    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n",
    "  elif color == \"rainbow\":\n",
    "    view.setStyle({'cartoon': {'color':'spectrum'}})\n",
    "  elif color == \"chain\":\n",
    "    chains = len(target_sequences) + 1\n",
    "    for n,chain,color in zip(range(chains),alphabet_list,pymol_color_list):\n",
    "       view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n",
    "\n",
    "  if show_sidechains:\n",
    "    BB = ['C','O','N']\n",
    "    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
    "                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n",
    "                        {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
    "                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "  if show_mainchains:\n",
    "    BB = ['C','O','N','CA']\n",
    "    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "\n",
    "  view.zoomTo()\n",
    "  return view\n",
    "\n",
    "def get_distance_predictions(results, interpolate=True):\n",
    "    bin_edges = results[\"distogram\"][\"bin_edges\"]\n",
    "    bin_edges = np.insert(bin_edges, 0, 0)\n",
    "\n",
    "    if interpolate:\n",
    "        distogram_softmax = softmax(results[\"distogram\"][\"logits\"], axis=2)\n",
    "        distance_predictions = np.sum(np.multiply(distogram_softmax, bin_edges), axis=2)\n",
    "    else:  # pick maximum probability distance instead\n",
    "        distogram_argmax = np.argmax(\n",
    "            results[\"distogram\"][\"logits\"][:, :, :63], axis=2\n",
    "        )  # skips last bin to avoid being too conservative\n",
    "        distance_predictions = bin_edges[distogram_argmax]\n",
    "\n",
    "    return distance_predictions\n",
    "\n",
    "def load_results(filepath, interpolate=True):\n",
    "\n",
    "    with open(filepath, \"rb\") as p:\n",
    "        results = pickle.load(p)\n",
    "        distance_predictions = get_distance_predictions(results, interpolate)\n",
    "        if \"predicted_aligned_error\" in results:\n",
    "            pae = results[\"predicted_aligned_error\"]\n",
    "        else:\n",
    "            pae = None\n",
    "    return distance_predictions, pae\n",
    "\n",
    "def plot_distances(filepath, distances, pae=None, limitA=None, limitB=None):\n",
    "\n",
    "    if pae is not None:\n",
    "        fig, ax = plt.subplots(1, 2)\n",
    "        d = ax[0].imshow(distances)\n",
    "        plt.colorbar(d, ax=ax[0], fraction=0.046, pad=0.04)\n",
    "        p = ax[1].imshow(pae)\n",
    "        plt.colorbar(p, ax=ax[1], fraction=0.046, pad=0.04)\n",
    "        ax[0].title.set_text(\"Distance map\")\n",
    "        ax[1].title.set_text(\"Predicted aligned error\")\n",
    "        ax[1].set_yticks([])\n",
    "    else:\n",
    "        fig, ax = plt.subplots()\n",
    "        d = ax.imshow(distances)\n",
    "        plt.colorbar(d, ax=ax, fraction=0.046, pad=0.04)\n",
    "        ax.title.set_text(\"Distance map\")\n",
    "\n",
    "    if limitA and limitB:\n",
    "        # plots a bounding box if any\n",
    "        rect1, rect2 = get_bounding_boxes(limitA, limitB)\n",
    "\n",
    "        if pae is not None:\n",
    "            rect3 = copy(rect1)\n",
    "            rect4 = copy(rect2)\n",
    "            ax[0].add_patch(rect1)\n",
    "            ax[0].add_patch(rect2)\n",
    "            ax[1].add_patch(rect3)\n",
    "            ax[1].add_patch(rect4)\n",
    "        else:\n",
    "            ax.add_patch(rect1)\n",
    "            ax.add_patch(rect2)\n",
    "    \n",
    "    plt.savefig(filepath, dpi=200, bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "def compare_to_native(\n",
    "    filepath,\n",
    "    pdb_path,\n",
    "    predicted_distances,\n",
    "    limitA=None,\n",
    "    limitB=None,\n",
    "):\n",
    "\n",
    "    parser = MMCIFParser(QUIET=True)\n",
    "    structure = parser.get_structure(\"model\", pdb_path)\n",
    "\n",
    "    compared_dist = np.triu(predicted_distances)\n",
    "\n",
    "    c_alphas = [r[\"CB\"] if \"CB\" in r else r[\"CA\"] for r in structure.get_residues()]\n",
    "\n",
    "    for i, ca_i in enumerate(c_alphas):\n",
    "        for j, ca_j in enumerate(c_alphas):\n",
    "            if j < i:\n",
    "                dist = ca_i - ca_j\n",
    "                compared_dist[i, j] = min(dist, 22)\n",
    "\n",
    "    if limitA and limitB:\n",
    "        predicted_distances = predicted_distances[\n",
    "            limitA[0] - 1 : limitA[1], limitB[0] - 1 : limitB[1]\n",
    "        ]\n",
    "        real_dist = np.zeros_like(predicted_distances)\n",
    "        real_dist = np.transpose(\n",
    "            compared_dist[limitB[0] - 1 : limitB[1], limitA[0] - 1 : limitA[1]]\n",
    "        )\n",
    "    else:\n",
    "        predicted_distances = np.triu(predicted_distances)\n",
    "        real_dist = np.transpose(np.tril(compared_dist))\n",
    "\n",
    "    fig, ax = plt.subplots(1, 2, gridspec_kw={\"width_ratios\": [1, 0.9]})\n",
    "    plt.subplots_adjust(\n",
    "        left=0.1, bottom=0.1, right=0.9, top=0.9, wspace=0.4, hspace=0.4\n",
    "    )\n",
    "\n",
    "    d = ax[0].imshow(compared_dist)\n",
    "    plt.colorbar(d, ax=ax[0], fraction=0.046, pad=0.04)\n",
    "    ax[0].set_ylabel(\"← Template distances\")\n",
    "    ax[0].set_xlabel(\"Predicted distances →\")\n",
    "    ax[0].xaxis.set_label_position(\"top\")\n",
    "    ax[0].set_yticks([])\n",
    "    ax[1].set_xlim(0, 22)\n",
    "    ax[1].set_ylim(0, 22)\n",
    "    ax[1].set_box_aspect(1)\n",
    "    ax[1].scatter(real_dist.reshape(-1), predicted_distances.reshape(-1))\n",
    "    corr = pearsonr(real_dist.reshape(-1), predicted_distances.reshape(-1))[0]\n",
    "    #print(f\"    Pearson correlation: {corr:.3f}\")\n",
    "    lims = [\n",
    "        np.min([ax[1].get_xlim(), ax[1].get_ylim()]),  # min of both axes\n",
    "        np.max([ax[1].get_xlim(), ax[1].get_ylim()]),  # max of both axes\n",
    "    ]\n",
    "    ax[1].plot(lims, lims, \"k-\", alpha=0.75, zorder=0)\n",
    "    ax[1].set_xlabel(\"PDB model distances\")\n",
    "    ax[1].set_ylabel(\"Distogram distances\")\n",
    "\n",
    "    if limitA and limitB:\n",
    "        # plots a bounding box if any\n",
    "        rect1, rect2 = get_bounding_boxes(limitA, limitB)\n",
    "        ax[0].add_patch(rect1)\n",
    "        ax[0].add_patch(rect2)\n",
    "\n",
    "    ax[0].set_title(\"Template-Prediction\\ncomparison\")\n",
    "    ax[1].set_title(f\"Distance agreement\\nPearson r: {corr:.3f}\")\n",
    "\n",
    "    plt.savefig(filepath, dpi=200, bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "rank_n = int(ranked_pdb.value.split(\"_\")[-1].split(\".\")[0])\n",
    "result_pickle = out_dir.joinpath(f\"result_{ranking['order'][rank_n]}.pkl\")\n",
    "result_png = out_dir.joinpath(f\"result_{ranking['order'][rank_n]}.pkl.png\")\n",
    "result_agreement_png = out_dir.joinpath(f\"result_{ranking['order'][rank_n]}.agreement.pkl.png\")\n",
    "\n",
    "if not result_png.is_file() or not result_agreement_png.is_file():\n",
    "    dist, pae = load_results(result_pickle)\n",
    "    plot_distances(result_png, dist, pae)\n",
    "    compare_to_native(result_agreement_png, template_mmcif_path, dist)\n",
    "else:\n",
    "    plt.axis('off')\n",
    "    distplot = mpimg.imread(result_png)\n",
    "    compplot = mpimg.imread(result_agreement_png)\n",
    "    plt.imshow(distplot)\n",
    "    plt.show()\n",
    "    plt.axis('off')\n",
    "    plt.imshow(compplot)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55998817",
   "metadata": {},
   "source": [
    "### <font color='#2684ff'>Structures</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2f489a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "prediction_pdb = f\"{out_dir}/{ranked_pdb.value}.pdb\"\n",
    "\n",
    "print(color.BOLD + f\"Prediction ({ranked_pdb.value}.pdb)\" + color.END)\n",
    "pred = show_pdb(prediction_pdb, \"pdb\", show_sidechains.value, False, col.value)\n",
    "pred.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa948223",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(color.BOLD + \"Template\" + color.END)\n",
    "template_pdb = f\"{mmcif_path}/0000.cif\"\n",
    "temp = show_pdb(template_pdb, \"cif\", show_sidechains.value, False, col.value)\n",
    "temp.show()"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaoAAAB2CAIAAAAWZkBBAAAACXBIWXMAAA7EAAAOxAGVKw4bAAATSElEQVR4nO3deXhM9/4H8Pc5s2/ZF0RiSYRGkBAhQoOW2ve1tLh0417L1UXpovS2v1/b2/22vyq/ulq9LaqqVBLEVqUikYQIQjbbZCIrk8kyM+f+cdLpZMGMjEzG+bwejyfzPd8z85l5nnnP95zvWZiq6hoQQojA6HSFrLNrIIQQ56D4I4QIFMUfIUSgKP4IIQJF8UcIESiKP0KIQFH8EUIESnzXHieSk1ugjnvDcZyzS3BtDMM4uwRCHCy6b18bezIufdgzxV8zUfwRwdLpCu8++mu1+OyjBLw3fPBxHEcJSATLVeOP4ziz2Ww2my0PnVuPy+FTj2VZlmUpAYkwuWT8WbLPaDTyCUjxZy8+9UQikeVvZ1dESEtzvfjjOI794H3Zm28yer2za3FVZpWqfNky/XMLJWKxXKEQi8WUgESAXGzqgx/3KXx9KPuaLz09Q6VSuWk0CqVSJBJRAhJBcbGpD47j+Pij7HOIGzdu1FRXSyQSiVTKMAzLsjQTQgTFleIPgMlkMhqNlofbtm6rNFTyX13YcxjHg7ev8K7vnf/xADB79my+pahIZzKZ1BqNUqFg/0AJSITDZeLPEljWyWWoMtTW1ir4vVcMy4cguR2z2QyrZDMYDIbKytraWpPZZAlHQoTDZeLPwvpbyjCMXC739vbRaNQSsUQsFltGgsSaZb+B0fTn2FkkElHqESFzpfjj/mBpYRhGIpFoNGovL2+ZTGaZwaSttwY4DhzHGY1G610HDMNYDh2nECQC5Erx1ySWYaUSqUwmk8tlErGEYWkTuAl8wJmMxlqr+KMPigic68cfy4pEIrFYLBFLxBIJS/F3GyaTiWVZDvVGeTToI0Lm8vEHgN/fx7B/oq3fBiwxJ2JFzq2EkNbjAYk/PgEtnF1Rq8P/HphMJlj9MNCPBBG4Byb+/vwb9MVuhB/9MTQrRIgVGigRQgSK4o8QIlCuEX80QUkIcTjXiD9CCHE4V4o/OjnhfrBMhdDHS4TmQZj5bay6unr7rviiGyUAp1Iqx40c7uvj5eyi7lF1Tc2qNf/7+oplGrXa2bUQ8kBpVvwdT05NTDrU5LEUXl6ez86bzV9LvYHvf/xZq9UteW5+c176zk6kpsmk0jkzpwA4f+FS/L6kJ2ZMsWXFjMxzifsPFhWXeLhp4gbFDIiOukPnxAOHY6OjVCqlXbU1udaK1W9bn40bFxsT1i1kd8L+xc/ex0+JEIFrVvydPZc9/4mZAe3aNF70ybqNFTdveXq4N2ivqa09nZnFgNEWFrXx923Oq9+BwVAd0LaNu5sGQHhYaNKRo6vf/sCy1N/PZ/4T06VSaYO1cvMKtvy4c9LYUV1DOl25rv1u206FTB7ZK/x2r7LvwJGI8O72xt/t1lr8zLygwPaWhzU1NZPGjbbrmQkhdmlW/HHgcJvDaG93dO3pM1n+fn4Bbf2TU9PGjhzGN65958OwbqG5eQWVBkP/vr2HD41r0NgvKpJvvAdKpXLFskXWLdt2/pJ5LjuyZ/cGPRMPHBk8MKZ3r3AAXUOCxzz2SML+g5G9wo+dSDlz9vxTcx8HcCoj89DRYzMnT1i38Ruz2fzZ+k3jRw+vqan5PeWUXCorKi7WaNQzJ0/w9fG661qNC7B2MTfv5z37Xlq60LpRqyvaumOXVqsLCGg7Y9I4L0+PysrK77f/fDE3XyGXDR8SFx0VYfsnQ3v6iMC19NTHiVPpkT3CIsLDUtJOm0wmS7tUIl3+t2eWPLfg2InU8xcvNWg8btVor+qamqIbxbqiP/+JGLa4pKS4tKzB9/+aVtulcyfLw+DOHW+UlNbUNHEvFH8/n1dfXMqy7MIFT/IpVlxSOmn8qFXPLw4N7vyfH35qspLGa9mlprb2y42be/cKf2Pl8pDOHflXSTxwRCIRr1m5/MmZU3/cHV9cWmbv0xIiWC069VFSWpabVzB76gS1Ws2yTNaFS+EPhfKLwrqFMAzj7qaJ7Nk9M+tC15Bg68aInt0zsy5YZ5ON9uw9kH46083drUH7dR1OZ2Z16tBhwpjH+BaO4/T6SuttUpVSAeCmvtKWF/L38/Xx8gQQN7D/voNHbtl/N5KPv/iK/0Mqlb712kuNO1y4mMOybGy/vgAGD4zZm3TYUGkQi0T6yqqy8oqg9u1WLFukVCrsfV1CBKtZ8cfAvhNIT6amd+oYpNFoAPQK734y9ZQl/izc3NT5BVdtabyrxKTDZeXlLy5d2ORFEPbsPcCZ6105VaVS6fV6H++6OWJ9pQGAxs5dewq5XCqR3Lxld/w12PfXWHlFRXnFzVfffJd/KBaJym/dGv5IXNKho+s2bhaJ2CEDB/TtY8fGLyEC16z4Cwnu+OXGb0WiJkJQo1a5aeodqMFxXPKp9NKy8udfWcu3sCx7S69Xq1TW3UpKyzWahkd4NNl4Z2kZmSbO/MzcWU1m38lTGTn5Bc/Om23dGNDGPzsnr0NQIP8wJzff28tTKpWKRSL+Zuq20Ov1NbW1GrXKrrVs4aZxa+vvu2zR09aNV65phwyKGfHo4Gvawv/b8LW/n29QYIADX5SQB9g9xl9ufkFu/mUAD8dG367PoaPHpRLJwJi6Dhdz8/T6yrWrXlAo5HzLux9/npJ2Oi62P4Dk1PSAtm2LS0rTMjJnz5jEd7A0pp/OnDV9kh3vSizS3bjx/F+fafLIm6zz2UmHf120YG6DpY898vCXm/7j7eUVGtzxqrZwV8K+0cMfAeDj7XVVqy0pLQNwLDnF0l8qlZRXlPt4ewK4dl2bm1cQENA2Ielwh6BAtUp117XsvTBXl84df/jpl+PJqdF9InRFxUd/T540duQvCfvb+PuNGj6Ev1sbzWYQYrt7jL/yilv7D/5abTUt4O6mcXd3L7h8xdLCMExocGdL/CWnpPWO6GHJPgCx/fv+9vtJPv5EItE/P/2iptY4cEB0aHBnvoOlMTYmOjS4s+2Dqcie4alpp9/79Isml/r5+Dw1Z1bjQ086BAVOnzQ+Yf+B7374yd1NMzQutk9kTwCdOgZF9Oj+z0/XqZWKkOBOV69r+f4DoqP+/5stk8eOBODp7n7gyLGc/Hw/X5/Hp060Za2o3r1sfDs8uVz21NzHd+yK/zl+n1qlHPHoEIZhpk4cs3XH7tX/875ELBk0ILpD0J02nwkh1piq6iZmNlvY2nc+fHzqhOBOHW/XyA9qzGaz0Wisra31/mP33ObNmxVyRUhIiK+fr1yukEgkIpGo5a9q9/vJUynppxfOf7IlX9Qudbd5a5WfHiFOodMVutI5v8SxKOaIwFH8EUIEqlVc8uDVF5fa2Ng69YuK7BcV6ewqCCH2odEfIUSgXCn+aK/8/UBHyhDBcqX4I4QQB3KN+KNBXwugwTURGteIP0IIcTiKP0KIQLWKA18ciOM4hmHo1FdCyF09CKM/s9nMcXU3KqPblTXJbDbzZ0w79iI0hLi0B2H0x5/QajKZ+Cu48ANAZxfVulg+Is4q/uh3ggicy8cffya/0WgUiVgA/En7zi6qNTKbzWaTqdbqfnKgKXUibK4df7NmzXJ2CS6MNoSJwLnSvj/mD+b6F4gm98Aol9NeAiJwrhR/PIZhypctMyvtuwUHsWaUy1PHjbO+0x4hAtQqLndqI47j+N18er2+pKSkuLi4qEhnMBgYhrH3wvHCxM918NMgDMNIpVJPT6+goCBvLy+5QiEWi/mPkYaERAh0ukIX2/fHz+3KpFK1Wm02mTiOqzIYYPXFxt0mNB/U77a907isSCSTydzd3BQKhUQqZVmWso8IjSvFH//NZFlWIpWqVEqO4xiWra6q4o/5g23xJ2T8B8j/L2JFEqlUqVQqFHKRSETZRwTIleIPAL+dKxaLATnLsDKZzGg0Wg51puCzhWUGSSQSSSRiubxus5eyjwiNi8UfAP57y/8vlkj4ozco+OzFJyD/W0J7TokwuV784Y8xIACxuK5+Gv3ZznoTmP8VAW32EkFyyfjj8dtr1pFH8WcL66SzjkJChObu8XciObkF6iCEEIeI7tvXxp6udNwfIYQ4Ct3mnBAiXBR/hBCBovgjhAgUxR8hRKAo/gghAkXxRwgRKIo/QohAUfwRQgSK4o8QIlAUf4QQgaL4I4QIFMUfIUSgKP4IIQLlwtf7I4TcD/oaTPu3+HKZ610FMtCD2zLHqJLa2p9Gf4SQeuqyz+WuHczhchkz7d92DOko/ggh9dRln8sN/pi6BLR9DYo/QkgjLpd9PDvLpvgjhAiUk6c+xn8/7l+jPnv76FujQ0Zfqbii0+tGhIyMahvl3KoIIULgzPjT6q+naFP25+zLKMzwkHkU6gtLDSVt1G0o/gghLcCRtzqqMdXklOXcrL7ZzaebRqq5c+c1h9/Yfu4HP7V/g3aT2SRmxFunbHOTuTmqMEKI7Xq+K3F2Cc2S8UKtLd10ukIHjP6Sr5346MSHOaU5apkmxDPEQ+ZxpuhMWXWZSqx6qvfTE7tNZBrtkMwrz9uWtfXYX35XSVSNn3BZ4tLNZ755rs/C5tdGCGlVOngicSG6/sPZdQBo5sbv+eLzK5NWdPEKXR23JsQzpMHS0qrSH89tH/r1kCXRSyd0nWBpj1rf+0zRmXGh41QSlZkzf3TiQ32tnl80p+fcQLfA6IB+83bO+XviMsPL1c0pjxByv7Vzx5qR6NEOhy9hxU6YrI4WFDFYOhjjwuGuQHYR3k3CiXznFdqUZsVfQUX+3/svjw0cCCC7JHtvbmJJZQkAmVgWExAzMGjQXyLmjwkds+XsFuu1No7fdPzKscScBACfpfxr98Xdg4MGAyisLFzw8/yEWYkApodNnx+5oDm1EUJawJju0N7E859j3XT0CaoXcAti8HAwntwM3U3EheDzqRj75b28hIipl6oO1Kz4G9ZpOIAzRWdeObCqoqZifOj4MN/uGpk6rzRv0+lNSxOXvDhgxbSwaX+N+pv1Wu8de+dC8YX2bu0B6G7p5vSY80TPJwHoa/X9NtTdnv23K79pbxU+PCuuOeURQu4rTyXGhSPjGsoMuFKO9u44YbX0oTZIykZ+CQDEZ+EhfwS4Q3cTAOb3x7x+ALA2EQlZABAdhFceQ3sPZBfhld3ILsLUCMyOgqEWUjEmbcBj3bB8KDzkOHgRr/2CKqMD6nfAvr+8srxFfRfxUVinE57C03nleT+d39G4/4D2sSqpukivu8NztncLHNRhUPNrI4Q4UEcvzOgNHxV+SEdXfyyMReoVDOuKnu3gJsfqPfU6787EP0bjajnis3CrGh8cBIAOngDgrsCwzzA2HKtHICELSik+m4al23EsF6uG47UReOJrAAj0xLxvceY6wtrg7bGY/y3O6fDxZDwTi48OOeDtOCD+xnQZAyBDl/H+8fd+Lfi1rLosul2/ub3mTQubtiR6aeP+HDiOqxvLSsXSIkMR/3dRZZFComh+PYSQ++SlR1FQgqO52DATu89i6lfIL4WYRWcfXGg0ntl7HiWVWBCDlcNw+BI+OoTc4rpF7x8AgMRzWDsKSimMJszYiIs3ACD+HIZ3q+uWqUX6VQCY2AMJ53DqKgBsSsaLQ1tN/PGuVFwe2vGRl2NX+Sl9j187nnUj63Y9GTAMUzcXPD1sxpKExfGX4vmHy/u/4Kh6CCEOF+yDTw4jS4vXR+C7VOSXAoDR3ET28VIuI+Uy3BWYEYlt8zB+fb2lZg4AxCwqa9DFF2tHw0MBuQRMo3PX/DUY3AXDugIAA1Q66Gg9h8XfqJDRAJ7fu/x88fkN4zaMDB51u54x7Qd4KbzXnfoCQBevLr/M3NOgQ2lVSS+/XnFBtOOPkNbl+1R89ThuVeOGHhtnYXcmNhzH4JC6md9tafU6vz8Rn/+K7CKUG/DFbxgSikHB+C2niaft4os3RmLqRuSXoE8gPp7csEPhLXyXirf2OvjtOPisjyCPoNyyHI30Tkcs9/Dr0cOvx7dnNq85/MbgjoMbLL1cfnlj2le7Zu72UzY8IpoQ4lwbjmNrGtzluFyGAHcsGoSfFiC3BB8fwopHcfoazlsNAxng5WFYE49r5Yhsj1BfZN9mkOihAMtCwsJXjSkREDUa/e3KxPoZ2HMWGdfxcDCCvbH+uAPejoPjb3HfJYv7LrGl59Yp23ac3zFl2+QZ3Wcm5e7v5tOtvLqislbft230npnx3kofxxZGCHGIiipUVAHA1XKs3AWVFGe1SDiHwV0QEVAv/lbuwpI4bJoNTyUKSvFmAk5erpv6aCC5ANvT8d1clBmw5RR6t4ei/okn6Vfx+h68NRZt3XBBh9cbbjHeI0ee9HYPOn/SceeMXauSVk7qNqmgokCr144OGc1vRxNCnMKuk95GheH5oTiQjQk9MH49rpTdv7psZftJb06OP0JIa2PvOb8RAQhvi8OXUFB6nyqyT4ue80sIEbK0q0i76uwi7gld7pQQIlAUf4QQgaL4I4QIFMUfIUSgKP4IIY243E1+eXaWTfFHCKkn0IPj75nrYjiAQaCHHXVT/BFC6tkyx1iXgK6FQaAHt2WOHRcCpMOeCSFCpNMV0uiPECJQFH+EEIGi+COECBTFHyFEoCj+CCECRfFHCBEoij9CiEBR/BFCBIrijxAiUBR/hBCB+i95/0LOfrZNFAAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "id": "cb638d00",
   "metadata": {},
   "source": [
    "## <font color='#2684ff'>Download results </font>\n",
    "\n",
    "Click on the button at the bottom of the menu:\n",
    "\n",
    "![image.png](attachment:image.png)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
